classes/User/User.mjs

```diff
@@ -0,0 +1,110 @@
+import dbDriver from "../../database/driver.mjs"
+import {
+  includeOnly,
+  removeProperties
+} from "../../utilities/removeProperties.mjs"
+
+export class User {
+  constructor(user = null, database = "mongo") {
+    this.database = new dbDriver(database)
+
+    if (user instanceof Object) {
+      this.id = user._id
+    } else {
+      this.id = user
+    }
+    this._fetchUser()
+  }
+
+  async _fetchUser() {
+    let user = await this.database.find({
+      _id: this.id,
+      "@type": "users"
+    })
+    if (Array.isArray(user)) {
+      user = user[0]
+    }
+    this.id = user?._id 
+    return (this.user = user)
+  }
+
+  async getUserById() {
+    // returns user's public info
+
+    return includeOnly(this.user, "profile")
+    // return removeProperties(this.user, "profile", "mbox", "cardInfo")
+  }
+
+  async getSelf() {
+    // returns full user object, only use this when the user is unthenticated i.e, logged in and getting himself.
+    return this.user
+
+    // const user = await this.database.find({
+    //   _id: userId,
+    //   "@type": "UserPreferences"
+    // })
+    // if (Array.isArray(user)) {
+    //   return user[0]
+    // } else {
+    //   return user
+    // }
+  }
+
+  async updateRecord(data) {
+    // updates user object. use with PUT or PATCH
+    if (!data) return
+    const previousUser = this.user
+    const newRecord = {...previousUser, ...data}
+    const updatedUser = await this.database.update(newRecord)
+    return updatedUser
+  }
+
+  async getProjects() {
+    const projects = await this.database.find({
+      "@type": "Project",
+      creator: this.id
```

Comment:
This would be only the ones that they have created. This needs to be able to find all projects with this id in the group. Right now, a group looks like

```json
{
  "_id": {
    "$oid": "65f48c9a4da4e4b8c656d0d2"
  },
  "slug": "exampleGroup",
  "members": [
    {
      "agent": "https://store.rerum.io/v1/id/5bec93c7d5deb589d44ec1a8",
      "roles": [
        "OWNER",
        "LEADER"
      ]
    }
  ],
  "custom_roles": [
    {
      "OWNER": {
        "label": "Owner",
        "permissions": "ALLOW_ALL"
      },
      "LEADER": {
        "label": "Group Leader",
        "permissions": "ALLOW_MODIFY_MEMBERS"
      }
    }
  ],
  "custom_defaults": {},
  "tags": [
    "GoG"
  ]
}
```

but maybe we should be using the `_id` since it is internal and have `id` instead of `agent`. That would allow an id map in `members` instead, which would be an easy lookup in mongodb `{'members.hexString':{"$exists":true}`

just thinking out loud

Reply:
agent is easy to get quickly from the `users` collection and could be left embedded even.